marimo/_runtime/executor.py

```diff
@@ -95,6 +95,37 @@ async def execute_cell_async(
         pass
 
 
+def handle_execute_cell_exception(
+    e: Exception, cell: CellImpl, glbls: dict[str, Any]
+) -> None:
+    tb = e.__traceback__
+    while tb and tb.tb_next:
+        tb = tb.tb_next
+
+    if tb is None:
+        raise e
+
+    filename = f"__marimo__cell_{cell.cell_id}_.py"
+
+    if filename in tb.tb_frame.f_code.co_filename and cell._start_line:
+        cell_lineno = tb.tb_frame.f_lineno
+
+        cell_lines = cell.code.split("\n")
+        offending_code = cell_lines[cell_lineno - 1]
+
+        file_path = glbls["__file__"]
+        source_lineno = cell._start_line + cell_lineno - 1
+        underline_str = "^" * len(offending_code)
+
+        e.add_note(
+            f"""  File "{file_path}", line {source_lineno}"""
+            + f"\n    {offending_code}"
+            + f"\n    {underline_str}"
+        )
```

Comment:
We need to be compatible back to Python 3.8, but unfortunately `add_note` was added in 3.11, so we can't use it.